llvm/docs/NVPTXUsage.rst

```diff
@@ -287,6 +287,62 @@ The ``@llvm.nvvm.fence.proxy.tensormap_generic.*`` is a uni-directional fence us
 
 The address operand ``addr`` and the operand ``size`` together specify the memory range ``[addr, addr+size)`` on which the ordering guarantees on the memory accesses across the proxies is to be provided. The only supported value for the ``size`` operand is ``128`` and must be an immediate. Generic Addressing is used unconditionally, and the address specified by the operand addr must fall within the ``.global`` state space. Otherwise, the behavior is undefined. For more information, see `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-membar>`_.
 
+Arithmetic Intrinsics
+---------------------
+
+'``llvm.nvvm.idp2a``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+.. code-block:: llvm
+
+    declare i32 @llvm.nvvm.idp2a(i32 %a, i1 immarg %a.unsigned, i32 %b, i1 immarg %b.unsigned, i1 immarg %is.hi, i32 %c)
+
+Overview:
+"""""""""
+
+The '``llvm.nvvm.idp2a``' intrinsic performs a 2-element vector dot product
+followed by addition. It corresponds directly to the ``dp2a`` PTX instruction.
+
+Semantics:
+""""""""""
+
+The 32-bit value in ``%a`` is broken into 2 16-bit values which are either sign
+or zero extended, depending on the value of ``%a.unsigned``, to 32 bits. Two
+bytes are selected from ``%b``, if ``%is.hi`` is true, the most significant
+bytes are selected, otherwise the least significant bytes are selected. These
+bytes are each sign or zero extended, depending on ``%b.unsigned``. The dot
+product of these 2-element vectors is added to ``%c`` to produce the return.
```

Comment:
I'm not convinced that passing %a.unsigned/%b.unsigned arguments to determine types of the other arguments is the best approach here. For me types should be properties of the function signature. Arguments affecting type interpretation are also more error prone (e.g. out-of range values, variables instead of immediates).

To me it looks like we may be better off having per-type intrinsics. `llvm.nvvm.idp2a.[su].[su]`. Then they will actually directly map to the corresponding `d2pa` instruction variants.

Reply:
In this case since the arguments are `i1`'s out-of range values should not be a concern. I've tagged the appropriate args with the `immarg` attr so variables instead of immediates should not be a problem either. It seems like here the decision is mostly a matter of stylistic preference with no concrete tradeoffs. I'd lean towards what I already have since this is what has been implemented internally and switching over will be somewhat painful. That being said, if you strongly prefer encoding this info in the name I can make the change. What do you think?

Reply:
It's indeed largely a naming problem in this case.

> this is what has been implemented internally and switching over will be somewhat painful.

Existence of non-public downstream implementation is a relatively weak argument for what's the right thing to do in LLVM itself. Considering that non-public changes did not go through the normal LLVM review process, some amount of changes is inevitable when the changes finally do face the public review. This is business as usual, IMO, and the maintenance price one pays for a diverged fork.

You may have existing users of your internal APIs that may be cumbersome to change. For that, the intrinsic auto-upgrade process may be helpful. It could be used to automatically translate internal intrinsic calls to the LLVM's.

https://github.com/llvm/llvm-project/blob/b368404dee8c341dc022a9e9a868f5a268e92033/llvm/lib/IR/AutoUpgrade.cpp#L10

Reply:
Fair enough, I understand it's not a strong argument but wasn't sure if this was a marginal case. I've updated the change to encode the type info in the intrinsic name.

Reply:
I would like to bring up one more point and would like to hear your thoughts, @Artem-B . I understand signedness of an intrinsic _operation_ should be encoded in the intrinsic name. However, when there are multiple operands with different signedness, does it scale to encode all combinations in the intrinsic name? This one expands to 4 intrinsic combinations, in the general case this could be more. So, I would like to understand what you think about this design given this general issue. Specifically, if we had an intrinsic with, say, 4 operands and signedness, would you still recommend to encode all combinations in the name?

Reply:
> However, when there are multiple operands with different signedness, does it scale to encode all combinations in the intrinsic name? 

It depends, I guess. Naming is hard, Good/consistent API design complicates things further. I do not think we have any standard set of rules on that. Yes, it's possible that if there are more type variants we may end up with too many intrinsics, or would have to resort to other options. If you have a specific example in mind, we could discuss how it could be handled.

> 4 operands and signedness, would you still recommend to encode all combinations in the name?

If it could all be offloaded to tablegen w/o complicating things, then my answer would probably be "yes". It does not mean that it's the final word on the matter. Other LLVM developers may have different opinions.

If there is existing code in LLVM using a different approach, I would consider the design choices and the arguments raised during review of that code. 

You may also ask on LLVM's discourse and see if wider LLVM community would offer any suggestions.

Reply:
Yes, keeping consistency with this suggested design is making me think twice on this. My question was a general one - if we or others see a similar need on > 2 operands how would we design them? In this context, I am also thinking about the design beyond their ease of implementation - does the replication of a particular intrinsic operation beyond a certain number seem reasonable for this operand signedness representation or not?

I also noticed another issue - sometimes we also need to express signedness on the return type. I cannot share the particular intrinsic yet, unfortunately. The idp* intrinsics also have a return type but we don't represent their signedness. However, on some others we might need them on the destination also.

This makes me wonder if the earlier design by @AlexMaclean keeps things more explicit and clear.

Would love your thoughts here, but these factors are keeping me inclined towards the previous design. :)

Reply:
> if we or others see a similar need on > 2 operands how would we design them?

One way to think on this is that most of the intrinsics fall into two categories:
- they provide a well defined functionality (e.g. math functions)
- they map to specific instructions.

The defined-functionality intrinsics usually define the intrinsic signature, so there's usually not much wiggle room.

The per-instruction intrinsics are probably better served by direct 1:1 intrinsic:instruction map, with intrinsic name roughly matching the instruction syntax. It works OK for 'normal' targets, but PTX is... unusual.

1:1 mapping is usually easily handled by tablegen, and LLVM can deal with a relatively large number of variants. LLVM already has ~11K intrinsics, and we already have multi-parameter variants for various forms of *MMA instructions that have a lot of different variants to accommodate different geometries and types. So, if you have only a handful of variables, 1:1 map is fine. 

If there's a good reason to push instruction variant selection for the sake of simplifying instruction variant selection, it should be discussed on a case by case basis. Often it does not provide as much benefit as it may appear. 

In cases like `dp*` instruction here we end up providing N extra immediate-only arguments, and it does not really save us anything -- instead of explicitly picking among N intrinsic names directly mapped to an instruction, we'd be picking a single intrinsic, and then have to choose from N instruction variants anyways. That just adds yet another indirection layer for the end user to be aware of.

Reply:
Another argument for per-instruction variant intrinsics is that for many PTX instructions not all instruction variants are valid. For intrinsics parametrized by arguments, if the user attempts to use an invalid variant, it will only be detected at the lowering phase, and will look like an LLVM crash. The errors in per-instruction-variant intrinsics will be obvious up-front and easy to understand/diagnose.

Reply:
FWIW, I think you meant that for N extra immediate-only arguments, we will instead need N^2 intrinsic names. But, irrespective of that, I see your points. Now I am more convinced than before about your suggestion. So, I am okay with the new intrinsic design.

Reply:
> for N extra immediate-only arguments, we will instead need N^2 intrinsic names

Yes, though `N^2` scaling applies only if the each argument is boolean. In general case, each argument may have more than 2 values, and we're looking at a full cartesian product of all such arguments. That obviously would not scale if we have too many combinations, but in practice it works well enough in most cases. We'll discuss/address exceptions on a case-by-case basis.